Issue #203 - Protecting protocol maps against concurrent modifications.

andrewkroh opened the pull request:
~~The pull request adds a mutex to protect the map containing `TcpSteam` data.~~ After this fix, Packetbeat should be safe to use when compiled with golang 1.5+ which changed the default GOMAXPROCS value to the number of logical CPUs instead on 1.

tsg commented on protos/tcp/tcp.go:
```diff
@@ -135,7 +162,7 @@ func (tcp *Tcp) Process(tcphdr *layers.TCP, pkt *protos.Packet) {
 			stream = &TcpStream{id: tcp.getId(), tuple: &pkt.Tuple, protocol: protocol, tcp: tcp}
 			stream.tcptuple = common.TcpTupleFromIpPort(stream.tuple, stream.id)
-			tcp.streamsMap[pkt.Tuple.Hashable()] = stream
+			tcp.putStream(pkt.Tuple.Hashable(), stream)
```
I think there's a theoretical race here: if two threads/goroutines do getStream at the same time and find that the stream doesn't yet exist, they can both call putStream() afterwards and overwrite each other's data. I think this cannot happen at the moment because all TCP packets are processed by a single goroutine anyway. Also, when we will scale to multiple threads, we're likely to want to send all packets from the same TCP stream to a single thread to avoid issues like this by design.

So I'm not saying that we necessarily have to change this, but we might want to be defensive here nevertheless. I'm curious what the others think as well.

andrewkroh replied:
I agree, there is definitely a race here (and in the protocol plugins) that would affect the correctness of this code if/when we parallelize the processing of incoming packets as described in #203. This change does not fix that. In fact running the parallel benchmark test with the race detector (`-race`) points out one of them.

I can open a separate issue to track the parallelization enhancement so we don't lose track of some of the comments regarding parallel processing from this issue.

tsg commented:
This should do it for the TCP map, but other protocols use a `transactionMap` in a similar non-thread-safe way. For example HTTP: https://github.com/elastic/packetbeat/blob/master/protos/http/http.go#L807

andrewkroh commented:
I will go through each of the protocol plugins and make sure they get similar updates. I should probably put some sort of concurrent map into libbeat common to avoid the code duplication.

tsg commented:
+1 for having a concurrent map in libbeat.

andrewkroh commented:
The `Cache` from elastic/libbeat#72 is included in this PR without doing a `godeps update github.com/elastic/libeat` because I didn't want to pull in other changes from libbeat that are not quite ready to be included.

This PR addresses the immediate needs for elastic/packetbeat#203 but the final solution will be to refactor Packetbeat to not require locking. Also by using this cache, Packetbeat will not need to create a new `time.Timer` for every transaction. Instead, a single goroutine can be used to periodically clean up expired entries in an instance of the Cache.

tsg commented:
I think the `Dns.expireTransaction` method is no longer used and can be deleted.

tsg commented:
@andrewkroh you ended up not needing the `RemovalListener` feature from libbeat? I don't see it being used in Packetbeat.

tsg commented:
I went through the changes, LGTM, one minor comment left above. Nice work!

@andrewkroh Can you merge master into it, please, so we can merge the PR from the web interface?

andrewkroh commented:
I rebased the PR and removed `Dns.expireTransaction`.

@tsg You're right, I didn't actually use `RemovalListener`, but if a protocol plugin decides to publish the partial event that was expired then they would need this notification mechanism. I will create an issue for the TODOs I removed that were placeholders for publishing events on expiration.
